File: #0006_ZigZagConversion/convert.java
class Solution {
    public String convert(String s, int numRows) {
        char[] c = s.toCharArray();
        int len = c.length;
        StringBuilder[] sb = new StringBuilder[numRows];
        for (int i = 0; i < sb.length; i++)
            sb[i] = new StringBuilder();

        int index = 0;
        while (index < len) {
            for (int i = 0; i < numRows && index < len; i++) // vertically down
                sb[i].append(c[index++]);
            for (int i = numRows-2; i >= 1 && index < len; i--) // obliquely up
                sb[i].append(c[index++]);
        }
        for (int i = 1; i < sb.length; i++)
            sb[0].append(sb[i]);
        return sb[0].toString();
    }
}

File: #0006_ZigZagConversion/convert.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows==1){return s;}
        string ans;
        vector<string>rows;
        for(int i=0;i<numRows;++i){
            string ss;
            rows.push_back(ss);
        }
        int pointer;
        for(int i=0;i<s.size();++i){
            pointer=i%(2*numRows-2);
            if(pointer>=numRows){
                pointer=-1*pointer+2*numRows-2;
            }
            rows.at(pointer).push_back(s.at(i));
        }
        for(int i=0;i<numRows;++i){
            ans+=rows.at(i);
        }
        return ans;
    }
};